### PR TITLE
Implement random method for LKJCorr

### PR DESCRIPTION
**Pull request by junpenglao**
using the algorithm in LKJ 2009(vine method based on a C-vine)

- [x] implement test.

### CODE REVIEW

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -1017,6 +1017,31 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
         self.tri_index[np.triu_indices(n, k=1)] = np.arange(n_elem)
         self.tri_index[np.triu_indices(n, k=1)[::-1]] = np.arange(n_elem)
 
+    def _random(self, n, eta, size=None):
+        beta = eta + (n-1)/2
+        P = np.zeros((n, n)) # partial correlations
+        r_triu = []
+        
+        for k in range(n-1):
```
Comment (ferrine) [inline pymc3/distributions/multivariate.py]: Is it possible to vectorize it?

Comment (junpenglao) [reply pymc3/distributions/multivariate.py]: I can not see an easy way to do it, the step below converting the partial correlation to raw correlation is especially tricky.

**Comment by junpenglao** [comment]
Any suggestion of implementing a test? I am stuck. @ferrine @aseyboldt

**Comment by ferrine** [comment]
What about comparing kde with theoretical density?

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -1017,6 +1017,31 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
...
 
+    def _random(self, n, eta, size=None):
+        beta = eta + (n-1)/2
+        P = np.zeros((n, n))  # partial correlations
+        r_triu = []
+
+        for k in range(n-1):
+            beta -= 1/2
+            for i in range(k+1, n):
+                P[k, i] = stats.beta.rvs(a=eta, b=eta)  # sampling from beta
```
Comment (ferrine) [inline pymc3/distributions/multivariate.py]: You'd better create these samples beforehand

**Comment by twiecki** [comment]
Or you could run a KS-test between true and sampled density.

**Comment by junpenglao** [comment]
I see - is there any similar test I can reference to in our test? @ferrine @twiecki

**Comment by aseyboldt** [comment]
This is similar I think: https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/sampler_fixtures.py#L28

**Comment by junpenglao** [comment]
Thanks everyone! I think I finally figure it out ;-)

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -1002,20 +1002,55 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
...
+        P = np.ones((n, n))
+        P[triu_ind] = stats.beta.rvs(a=beta, b=beta)
+        # scale partial correlation matrix to [-1, 1]
+        P = (P-.5)*2
+        r_triu = []
+
+        for k, i in zip(triu_ind[0], triu_ind[1]):
+            p = P[k, i]
+            for l in range(k-1, -1, -1):  # convert partial correlation to raw correlation
+                p = p * np.sqrt((1-P[l, i]**2)*(1-P[l, k]**2)) + P[l, i]*P[l, k]
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: pep8 wants white spaces around math operators

File: pymc3/distributions/multivariate.py [hunk 4]
```diff
@@ -1002,20 +1002,55 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
...
+
+        return np.asarray(r_triu)
+
+    def random(self, point=None, size=None):
+        n, eta = draw_values([self.n, self.eta], point=point)
+        if n < 3:
+            samples = generate_samples(stats.beta.rvs, eta, eta,
+                                       dist_shape=self.shape,
+                                       size=size)
+            samples = (samples-0.5)*2
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: white-spaces

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -1002,20 +1002,55 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
...
+        P = np.ones((n, n))
+        P[triu_ind] = stats.beta.rvs(a=beta, b=beta)
+        # scale partial correlation matrix to [-1, 1]
+        P = (P - .5) * 2
+        r_triu = []
+
+        for k, i in zip(triu_ind[0], triu_ind[1]):
+            p = P[k, i]
+            for l in range(k-1, -1, -1):  # convert partial correlation to raw correlation
+                p = p * np.sqrt((1 - P[l, i]**2) *
```
Comment (junpenglao) [inline pymc3/distributions/multivariate.py]: what i am doing here is slow in a for loop, should I change it to `reduce` @ColCarroll?

Comment (aseyboldt) [reply pymc3/distributions/multivariate.py]: For stuff like this we could also consider to use numba if it is available. (check if it is available and create a no-op replacement if it is not).

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -1002,20 +1002,55 @@ def __init__(self, eta=None, n=None, p=None, transform='interval', *args, **kwar
...
         warnings.warn('Parameters in LKJCorr have been rename: shape parameter n -> eta '
                       'dimension parameter p -> n. Please double check your initialization.',
                       DeprecationWarning)
         self.tri_index = np.zeros([n, n], dtype='int32')
-        self.tri_index[np.triu_indices(n, k=1)] = np.arange(n_elem)
-        self.tri_index[np.triu_indices(n, k=1)[::-1]] = np.arange(n_elem)
+        self.tri_index[np.triu_indices(n, k=1)] = np.arange(shape)
+        self.tri_index[np.triu_indices(n, k=1)[::-1]] = np.arange(shape)
+
+    def _random(self, n, eta, size=None):
```
Comment (ferrine) [inline pymc3/distributions/multivariate.py:1020]: should be static I suppose

Comment (ferrine) [reply pymc3/distributions/multivariate.py:1020]: seems like size argument is ignored

Comment (junpenglao) [reply pymc3/distributions/multivariate.py:1020]: Yep. I ignored the size as the _random method here can only generate 1 slide of the random matrix.

Comment (junpenglao) [reply pymc3/distributions/multivariate.py:1020]: Thanks for the suggestion - I manage to use the size and vectorized the `_random` method.

**Comment by junpenglao** [comment]
OK the implementation should be correct as the distribution of the matrix elements match `beta()` distribution. However I am getting none positive-definite matrix when the dimension n>3:
```python
import numpy as np
from scipy import stats

# cherry picked LKJ random function
def lkj_random(n, eta, size=None):
    beta0 = eta - 1 + n/2
    shape = n * (n-1) // 2
    triu_ind = np.triu_indices(n, 1)
    beta = np.array([beta0 - k/2 for k in triu_ind[0]])
    # partial correlations sampled from beta dist.
    P = np.ones((n, n) + (size,))
    P[triu_ind] = stats.beta.rvs(a=beta, b=beta, size=(size,) + (shape,)).T
    # scale partial correlation matrix to [-1, 1]
    P = (P - .5) * 2
    
    for k, i in zip(triu_ind[0], triu_ind[1]):
        p = P[k, i]
        for l in range(k-1, -1, -1):  # convert partial correlation to raw correlation
            p = p * np.sqrt((1 - P[l, i]**2) *
                            (1 - P[l, k]**2)) + P[l, i] * P[l, k]
        P[k, i] = p
        P[i, k] = p

    return np.transpose(P, (2, 0 ,1))

def is_pos_def(A):
    if np.array_equal(A, A.T):
        try:
            np.linalg.cholesky(A)
            return 1
        except np.linalg.linalg.LinAlgError:
            return 0
    else:
        return 0

P = lkj_random(4, 1., 1000)
k=0
for i, p in enumerate(P):
    k+=is_pos_def(p)
print(k)
```

Thoughts? @aseyboldt

**Comment by aseyboldt** [comment]
@junpenglao I haven't look at it in detail, but should there really be that many 1. in the partial correlations `P`? Maybe something like `P = np.ones((n, n) + (size,)) / 2`?

**Comment by junpenglao** [comment]
@aseyboldt at the end only the diagonal is 1. Currently LKJ random returns the upper triangular elements (same as the distribution) but I was just doing some test in the code above.

**Comment by aseyboldt** [comment]
Sorry, you are right.
The partial correlations are just a scaled version of the precision matrix, right? So maybe you could use that to check the result?

**Comment by junpenglao** [comment]
@aseyboldt not sure I understand what you mean.

~~I also compare with a [Julia implementation](https://github.com/jmxpearson/VinDsl.jl/blob/master/src/distributions/LKJ.jl#L58-L76), both have the same output. So it might be a numerical stability issue.~~

[EDIT], trying to figure out whether I have the same output as Julia or Stan.

**Comment by aseyboldt** [comment]
hm. But shouldn't `A_{ij}/np.sqrt(A_ii * A_jj)` equal the original partial correlations, were A = \sigma^{-1}? It isn't. And sometimes the eigenvalues are definitely negative, much more so than could be explained by some simple rounding.

**Comment by aseyboldt** [comment]
I'm using this to compute the partial correlations:
```
tau = linalg.inv(P[0])

partial = tau.copy()
partial /= np.sqrt(np.diag(tau))[None, :]
partial /= np.sqrt(np.diag(tau))[:, None]
```
(And just added a print statement in the function to get the original values)

**Comment by junpenglao** [comment]
I will also compare with the implementation in R from @rmcelreath https://github.com/rmcelreath/rethinking/blob/master/R/distributions.r#L165-L184, I cannot really figure it out in Stan and Julia

**Comment by aseyboldt** [comment]
I just noticed that it seems to work out if I make all loops one shorter:
```
    for k, i in zip(triu_ind[0][:-1], triu_ind[1][:-1]):
        p = P[k, i]
        for l in range(k-1, 0, -1):  # convert partial correlation to raw correlation
            p = p * np.sqrt((1 - P[l, i]**2) *
                            (1 - P[l, k]**2)) + P[l, i] * P[l, k]
        P[k, i] = p
        P[i, k] = p
```
This was pure experimentation though, and might be complete nonsense. I can have a closer look tomorrow.

**Comment by aseyboldt** [comment]
Sorry, my last comment was wrong, I looked at the wrong array. I just deleted it.

**Comment by junpenglao** [comment]
Turns out the R implementation from @rmcelreath is the most stable - test pass locally and samples are mostly positive definite (failed sometimes when n is large and eta<<1, but much better than the previous implementation nonetheless).

**Comment by junpenglao** [comment]
@aseyboldt The current random method could potentially be extended for generating LKJCholskyCov as it produces a triangular matrix as an intermedia step.

**Pull request merged**